### PR TITLE
Add Organization Id to the APIM internal service REST APIs

### PR DESCRIPTION
**Pull request by Jayanie**
## Purpose

- This PR adds the support to filter details based on organization Id. 

- This is related to the feature Deploying API Gateways in Private Data planes. Details like Applications, Subscriptions etc will be pulled from Global Adaptors. These details will be partitioned by the organization. So GA will only be able to fetch details related to a particular organization from the API Publisher.
To support this, API Manager should allow filtering APIs, and artifacts based on the organizationId.

REST API calls supported:

1. /applications
2. /runtime-artifacts
3. /subscriptions
4. /application-key-mappings
5. /retrieve-api-artifacts
6. /runtime-metadata
7. /apis/deployed-revisions

## Issues

https://github.com/wso2-enterprise/choreo/issues/12761
https://github.com/wso2-enterprise/choreo/issues/12906

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java [hunk 1]
```diff
@@ -8577,14 +8577,14 @@ public APIRevision getAPIRevision(String revisionUUID) throws APIManagementExcep
     /**
      * Get the revision UUID from the Revision no and API UUID
      *
-     * @param revisionNum   revision number
-     * @param apiUUID       UUID of the API
+     * @param revisionNum revision number
+     * @param apiUUID     UUID of the API
      * @return UUID of the revision
```
Comment (shanakama) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIProviderImpl.java:8582]: missed doc comment for organization param. Check other places too

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/SubscriptionsApiServiceImpl.java [hunk 2]
```diff
@@ -56,10 +60,18 @@ public Response subscriptionsGet(String xWSO2Tenant, Integer apiId, Integer appI
...
+                    subscriptionValidationDAO.getAllSubscriptionsByOrganization(organization))).build();
+        } else if (StringUtils.isNotEmpty(organization) && organization.equalsIgnoreCase(APIConstants.ORG_ALL_QUERY_PARAM) &&
+                xWSO2Tenant.equalsIgnoreCase(MultitenantConstants.SUPER_TENANT_DOMAIN_NAME)) {
+            result = Response.ok().entity(SubscriptionValidationDataUtil.fromSubscriptionToSubscriptionListDTO(
+                    subscriptionValidationDAO.getAllSubscriptions())).build();
         } else if (StringUtils.isNotEmpty(xWSO2Tenant)) {
             result = Response.ok().entity(SubscriptionValidationDataUtil.fromSubscriptionToSubscriptionListDTO(
                     subscriptionValidationDAO.getAllSubscriptions(xWSO2Tenant))).build();
-        } else {
+        }  else {
```
Comment (ruks) [inline components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/SubscriptionsApiServiceImpl.java:74]: a extra space?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 3]
```diff
@@ -391,6 +391,59 @@ public List<APIRuntimeArtifactDto> retrieveGatewayArtifactsByAPIIDAndLabel(Strin
...
+                    apiRuntimeArtifactDto.setContext(resultSet.getString("CONTEXT"));
+                    InputStream artifact = resultSet.getBinaryStream("ARTIFACT");
+                    if (artifact != null) {
+                        byte[] artifactByte = APIMgtDBUtil.getBytesFromInputStream(artifact);
+                        try (InputStream newArtifact = new ByteArrayInputStream(artifactByte)) {
+                            apiRuntimeArtifactDto.setArtifact(newArtifact);
+                        } catch (IOException e) {
+                            // Do not handle the exception here since runtime artifacts are retrieved by API UUID
+                            // throw the exception here.
+                            handleException("Error occurred retrieving input stream from byte array.", e);
```
Comment (shanakama) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:433]: `Error occurred while retrieving input stream from byte array.`  ?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java [hunk 4]
```diff
@@ -391,6 +391,59 @@ public List<APIRuntimeArtifactDto> retrieveGatewayArtifactsByAPIIDAndLabel(Strin
...
+                preparedStatement.setString(index, label);
+                index++;
+            }
+            try (ResultSet resultSet = preparedStatement.executeQuery()) {
+                while (resultSet.next()) {
+                    APIRuntimeArtifactDto apiRuntimeArtifactDto = new APIRuntimeArtifactDto();
+                    apiRuntimeArtifactDto.setApiId(apiId);
+                    String label = resultSet.getString("LABEL");
+                    // Do not handle the exception here since runtime artifacts are retrieved by API UUID
+                    // throw the exception here.
```
Comment (shanakama) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/GatewayArtifactsMgtDAO.java:414]: is this required here ?

File: components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/SubscriptionsApiServiceImpl.java (hunk 2 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.internal.service/src/main/java/org/wso2/carbon/apimgt/internal/service/impl/SubscriptionsApiServiceImpl.java]: Fixed with 3525a574b6e752661df194d3e7d8614a60abfdfa

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 5]
```diff
@@ -16348,19 +16348,27 @@ public APIRevision getRevisionByRevisionUUID(String revisionUUID) throws APIMana
...
-     * @param revisionNum Revision number
-     * @param apiUUID     UUID of the API
+     * @param revisionNum   Revision number
+     * @param apiUUID       UUID of the API
+     * @param organization  organization ID of the API
      * @return UUID of the revision
      * @throws APIManagementException if an error occurs while retrieving revision details
      */
-    public String getRevisionUUID(String revisionNum, String apiUUID) throws APIManagementException {
+    public String getRevisionUUID(String revisionNum, String apiUUID, String organization) throws APIManagementException {
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:16357]: can we have a new method for this

**Review by tharindu1st — CHANGES_REQUESTED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 5 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:16357]: ack

Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: Fixed with 7ebec1dc831d3a958a6721530f4d83ca1961ea07

**Review by tharindu1st — APPROVED** [review]

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java [hunk 6]
```diff
@@ -16372,6 +16373,37 @@ public String getRevisionUUID(String revisionNum, String apiUUID) throws APIMana
...
+     * @param revisionNum   Revision number
+     * @param apiUUID       UUID of the API
+     * @param organization  organization ID of the API
+     * @return UUID of the revision
+     * @throws APIManagementException if an error occurs while retrieving revision details
+     */
+    public String getRevisionUUIDByOrganization(String revisionNum, String apiUUID, String organization) throws APIManagementException {
+
+        String revisionUUID = null;
+        log.info("JAYANIEEEEE:" +organization);
```
Comment (Jayanie) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java:16388]: Remove log

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java [hunk 7]
```diff
@@ -2839,6 +2884,21 @@ public class SQLConstants {
...
+                    "API_PROVIDER, AM_GW_PUBLISHED_API_DETAILS.API_NAME AS API_NAME,AM_GW_PUBLISHED_API_DETAILS.API_VERSION AS " +
+                    "API_VERSION," +
+                    "AM_GW_PUBLISHED_API_DETAILS.API_TYPE AS API_TYPE,AM_GW_API_ARTIFACTS.ARTIFACT AS ARTIFACT," +
+                    "AM_GW_API_DEPLOYMENTS.LABEL AS LABEL,AM_GW_API_DEPLOYMENTS.VHOST AS VHOST, " +
+                    "AM_API.CONTEXT AS CONTEXT FROM " +
+                    "AM_GW_PUBLISHED_API_DETAILS,AM_GW_API_ARTIFACTS,AM_GW_API_DEPLOYMENTS,AM_API WHERE " +
+                    "AM_GW_PUBLISHED_API_DETAILS.API_ID=AM_GW_API_DEPLOYMENTS.API_ID AND " +
+                    "AM_GW_API_ARTIFACTS.API_ID=AM_GW_API_DEPLOYMENTS.API_ID AND " +
+                    "AM_API.API_UUID=AM_GW_API_DEPLOYMENTS.API_ID AND " +
+                    "AM_GW_API_ARTIFACTS.REVISION_ID=AM_GW_API_DEPLOYMENTS.REVISION_ID AND ";
```
Comment (Jayanie) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java:2900]: Remove `And` at the end

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java [hunk 8]
```diff
@@ -3438,8 +3498,10 @@ public static class APIRevisionSqlConstants {
         public static final String GET_MOST_RECENT_REVISION_ID = "SELECT REVISIONS_CREATED FROM AM_API WHERE API_UUID" +
                 " = ?";
         public static final String GET_REVISION_BY_REVISION_UUID = "SELECT * FROM AM_REVISION WHERE REVISION_UUID = ?";
-        public static final String GET_REVISION_UUID = "SELECT REVISION_UUID FROM AM_REVISION WHERE API_UUID = ? " +
-                "AND ID = ?";
+        public static final String GET_REVISION_UUID = "SELECT REVISION_UUID FROM AM_REVISION, AM_API WHERE AM_REVISION.API_UUID = ? " +
```
Comment (Jayanie) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java:3501]: Remove `AM_API `

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/APIArtifactGeneratorUtil.java [hunk 9]
```diff
@@ -34,7 +34,7 @@ public class APIArtifactGeneratorUtil {
     private static final GatewayArtifactsMgtDAO gatewayArtifactsMgtDAO = GatewayArtifactsMgtDAO.getInstance();
 
     public static RuntimeArtifactDto generateAPIArtifact(List<String> apiUuids, String name, String version,
-                                                         String gatewayLabel, String type, String tenantDomain)
+                                                             String gatewayLabel, String type, String tenantDomain)
```
Comment (Jayanie) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/APIArtifactGeneratorUtil.java:37]: Rename tenant Domain to organization

**Comment by malinthaprasan** [comment]
Improve error handling in the PR as overall. Do not log and continue when the request should be failed.

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java (hunk 6 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: Fixed with c26ce6f96326ae105cf262d8972353dc1aca45c7

Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/ApiMgtDAO.java]: Fixed with 205b70c752921ba3104f71e0633cad76ab0c0da7

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java (hunk 7 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java]: Fixed with 0745fad4ee8b0890e2a4bd20d479acfd5876f6ce

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java (hunk 8 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/dao/constants/SQLConstants.java]: Fixed with 0745fad4ee8b0890e2a4bd20d479acfd5876f6ce

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/APIArtifactGeneratorUtil.java (hunk 9 above)
Comment (Jayanie) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/gatewayartifactsynchronizer/APIArtifactGeneratorUtil.java]: Fixed with 0745fad4ee8b0890e2a4bd20d479acfd5876f6ce

**Review by malinthaprasan — APPROVED** [review]

**Pull request merged**